pattoo_shared/installation/configure.py

```diff
@@ -0,0 +1,313 @@
+            None
+
+        """
+        # Initialize key variables
+        if server is False:
+            config_file = os.path.join(config_directory, 'pattoo.yaml')
+        else:
+            config_file = os.path.join(config_directory, 'pattoo_server.yaml')
+
+        # Set config_dict if None has been passed in
+        if config_dict is None:
+            if server is False:
+                config_dict = self.default_config
+            else:
+                config_dict = self.default_server_config
+
+        # Say what we are doing
+        print('\nConfiguring {} file.\n'.format(config_file))
+
+        # Get configuration
+        config = self.read_config(config_file, config_dict)
+
+        if server is False:
+            # Check validity of directories
+            for key, value in sorted(config['pattoo'].items()):
+                if 'directory' in key:
+                    if os.sep not in value:
+                        log.log2die_safe(
+                            1019, '{} is an invalid directory'.format(value))
+
+                    # Attempt to create directory
+                    full_directory = os.path.expanduser(value)
+                    if os.path.isdir(full_directory) is False:
+                        files.mkdir(full_directory)
+
+                        # Recursively set file ownership to pattoo user
+                        # and group
+                        shared.chown(full_directory)
+
+        # Write file
+        with open(config_file, 'w') as f_handle:
+            yaml.dump(config, f_handle, default_flow_style=False)
+
+    def read_config(self, filepath, default_config):
+        """Read configuration file and replace default values.
+
+        Args:
+            filepath: Name of configuration file
+            default_config: Default configuration dict
+
+        Returns:
+            config: Dict of configuration
+
+        """
+        # Convert config to yaml
+        default_config_string = yaml.dump(default_config)
+
+        # Read config
+        if os.path.isfile(filepath) is True:
+            with open(filepath, 'r') as f_handle:
+                yaml_string = (
+                    '{}\n{}'.format(default_config_string, f_handle.read()))
+                config = yaml.safe_load(yaml_string)
+        else:
+            config = default_config
+
+        return config
+
+    def secondary_key_check(self, config, primary, secondaries):
+        """Check secondary keys.
+
+        Args:
+            config: Configuration dict
+            primary: Primary key
+            secondaries: List of secondary keys
+
+        Returns:
+            None
+
+        """
+        # Check keys
+        for key in secondaries:
+            if key not in config[primary]:
+                log_message = ('''\
+Configuration file's "{}" section does not have a "{}" sub-section. \
+Please fix.'''.format(primary, key))
+                log.log2die_safe(1057, log_message)
+
+    def check_config(self):
```

Comment:
Ideally this function should have the content of the configuration files, the primary keys and secondary keys as arguments. It doesn't need to be a method. It won't scale as written for other agents.